src/backend/cdb/cdbmutate.c

```diff
@@ -955,13 +1060,21 @@ apply_motion_mutator(Node *node, ApplyMotionState *context)
 			else
 			{
 				/*
-				 * no motion nodes in child plan nodes - no need for
-				 * ExplicitRedistribute: restore flow
+				 * Restore flow if Explicit Redistribute Motion is not needed
 				 */
 				flow->req_move = MOVEMENT_NONE;
 				flow->flow_before_req_move = NULL;
 			}
 
+			/*
+			 * If we're here, it means we are directly under the ModifyTable
+			 * node. We are about to go to out of the recursion and go into
+			 * other subtree. So reset the state and continue checking in case
+			 * of another Explicit Redistribute Motion is needed.
+			 */
+			context->mtIsChecking = true;
+			context->needExplicitMotion = false;
```

Comment:
I suggest to place this line under `if` above, because it's redundant to set needExplicitMotion to false when needExplicitMotion is already false.